mirrord/agent/src/steal/http/hyper_handler/httpv1.rs

```diff
@@ -0,0 +1,194 @@
+use core::{fmt::Debug, future::Future, pin::Pin};
+use std::{net::SocketAddr, sync::Arc};
+
+use bytes::Bytes;
+use dashmap::DashMap;
+use fancy_regex::Regex;
+use http_body_util::Full;
+use hyper::{body::Incoming, client, header::UPGRADE, service::Service, Request, Response};
+use mirrord_protocol::{ConnectionId, Port, RequestId};
+use tokio::{
+    macros::support::poll_fn,
+    net::TcpStream,
+    sync::{mpsc::Sender, oneshot},
+};
+use tracing::error;
+
+use super::{header_matches, matched_request, prepare_response, HyperHandler, RawHyperConnection};
+use crate::{
+    steal::{http::error::HttpTrafficError, HandlerHttpRequest, MatchedHttpRequest},
+    util::ClientId,
+};
+
+#[derive(Debug)]
+pub(crate) struct HttpV1 {
+    upgrade_tx: Option<oneshot::Sender<RawHyperConnection>>,
+}
```

Comment:
I've expanding on part of this in another comment on `handshake`.

> What is the advantage of having HttpV1 and HttpV2 as independent types?

This is sorta of a pattern we can use in rust to get a compile-time guarantee that the methods we're using cannot fail (kinda).

A benefit against `enums` here is that you can't call the wrong method (well, you can try, but it'll be very explicit that you're doing the wrong thing) as the compiler will either say "Hey, there are 2 types that implement this, pick one", when we use this pattern as a sort of method overloading, or it'll say "Error, you're trying to call a method that doesn't exist for this type".

As hyper already has this separation of types on HTTP version, we can turn this into an advantage and get code that is more obvious than the alternative, for example:

- typed call

```rust
// This becomes the "source of truth" for the whole flow that we're starting here, from now on the operations
// are completely typed (in a very strict way).
match request.version {
    Http1 => request.handle::<HttpV1>(arg1, arg2),
    Http2 => request.handle::<HttpV2>(arg1),
}
```

- enum call

```rust
// The `handle` method will have to `match` on which HTTP version this is, and any other methods in this flow will
// also have to check if we're actually dealing with the HTTP version they care about. We've pushed the
// responsibility of identifying HTTP version to the runtime in (possibly) multiple places, instead of just 1.
request.handle(arg1, Some(arg2));
```

Reply:
Alright, _traitfied_ a bunch of this stuff. I'll resolve the other comments that are related to this suggestion.

> In my opinion there is way too much code duplication between HttpV1 and HttpV2 and I don't see the justification.

Most of it is now implemented inside `HttpV` traits, so the meaty duplication should have been lessened.